File: python/second/partition_equal_subset_sum.py
class Solution(object):
    # O(n * t) time | O(n * t) space
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        sums = sum(nums)
        if sums % 2 != 0:
            return False
        
        target = sums // 2
        dp = set()
        dp.add(0)

        for num in nums:
            newDp = set()
            for d in dp:
                t = num + d
                if t == target:
                    return True
                newDp.add(d)
                newDp.add(t)
            dp = newDp
        
        return False

solution = Solution()
nums = [1,5,11,5]
print(solution.canPartition(nums))

File: src/main/java/december_2025/PartitionEqualSubsetSum.java
package december_2025;

import java.util.*;

public class PartitionEqualSubsetSum {

    public static void main(String[] args) {
        int[] nums = {1, 5, 11, 5};
        PartitionEqualSubsetSum partitionEqualSubsetSum = new PartitionEqualSubsetSum();
        boolean result = partitionEqualSubsetSum.canPartition(nums);
        System.out.println(result);
    }

    // O(2^n) time | O(n) space
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % 2 != 0) {
            return false;
        }
        int target = sum / 2;

        Set<Integer> dp = new HashSet<>();
        dp.add(0);

        for (int num : nums) {
            Set<Integer> nextDp = new HashSet<>();
            for (int t : dp) {
                if (t + num == target) {
                    return true;
                }
                nextDp.add(t);
                nextDp.add(t + num);
            }
            dp = nextDp;
        }
        return false;
    }

    private boolean dfs(int[] nums, int i, int sum, int target, Map<String, Boolean> memo) {
        if (sum == target) {
            return true;
        }
        if (i == nums.length || sum > target) {
            return false;
        }
        String key = i + ":" + sum;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        memo.put(key, dfs(nums, i + 1, sum + nums[i], target, memo) || dfs(nums, i + 1, sum, target, memo));
        return memo.get(key);
    }

}
